### ISSUE TITLE
NamedQuery mangles query during processing.

### ISSUE THREAD

**Issue by teopapath** [spring-projects/spring-data-jpa#3085]
Issue related to https://github.com/spring-projects/spring-data-jpa/issues/2990.
Test case to replicate it is available in [error2990_2](https://github.com/teopapath/spring-boot-error/tree/master/src/main/java/com/example/demo/error2990_2)

**Comment by gregturn** [comment]
First of all, you can get things working if you simply move the original query straight to the repository method:

```java
    @Transactional(readOnly = true)
//    @Query(name =  Account.NAMED_QUERY_FIND_ACCOUNT_CONTACTS_BY_ACCOUNT_IDS_AND_CONTACT_TYPE)
    @Query("select new com.example.demo.error2990_2.AccountContactInfoDTO(acc.id, acc.name, VALUE(contacts)) "
            + "from Account acc "
            + "join acc.contacts contacts on KEY(contacts) = :contactType "
            + "where acc.id in (:accountIds)")
    List<AccountContactInfoDTO> findAccountContactsByAccountIdsAndContactType(List<Long> accountIds, AccountContactType contactType);
```

When you supply it via named query, then the original query is fed to Hibernate to pre-processing, and later extracted by `String queryString = extractor.extractQueryString(query)`, thus yielding:

```
select new com.example.demo.error2990_2.AccountContactInfoDTO(acc.id, acc.name, contacts.{element})
from com.example.demo.error2990_2.Account acc
join acc.contacts contacts
on contacts.{index} = :contactType
where acc.id in (:accountIds)
```

This is clearly wrong, and so when Spring Data JPA tries to move forward with it, it doesn't work as expected. It appears to be partial processing on Hibernate's behalf, make me wonder if there is a better place to do this whole process.

**Comment by gregturn** [comment]
And to be clear, #2990 was a problem due to our parser not handling `!=` in the extracted query, which was resolved.

**Comment by teopapath** [comment]
That was also the workoaround for 2990 and that’s why i related the 2 issues.

**Comment by gregturn** [comment]
Okay, I've traced things back to `HibernateUtils`, the utility used to get the HQL back from the `Query` object. Looks like there is more than one way to get the query, and sometimes one is right, and sometimes the other one is right. So I'm 

This drafted solution works with the query up above and passes our tests, but seems a bit hack, so I'm going to keep digging.

```java
	public static String getHibernateQuery(Object query) {

		try {

			// Try the new Hibernate implementation first
			if (query instanceof SqmQuery sqmQuery) {

				String hql = sqmQuery.getQueryString();

				if (!hql.equals("<criteria>")) {
					return hql;
				}

				String sqmHql = sqmQuery.getSqmStatement().toHqlString();

				return sqmHql;
			}

			// Couple of cases in which this still breaks, see HHH-15389
		} catch (RuntimeException o_O) {}

		// Try the old way, as it still works in some cases (haven't investigated in which exactly)

		if (query instanceof Query<?> hibernateQuery) {
			return hibernateQuery.getQueryString();
		} else {
			throw new IllegalArgumentException("Don't know how to extract the query string from " + query);
		}
	}
```

**Comment by teopapath** [comment]
@gregturn  Is there any update on this issue?

**Comment by gregturn** [comment]
@teopapath Not yet. I have drafted a solution that seems to work, but I'm not convinced we are covering all the corner cases. I'd like to get @mp911de 's feedback before moving forward.

**Comment by mp911de** [comment]
Let's give it a try.

**Comment by gregturn** [comment]
Merged to `main`, backported to `3.1.x` and `3.0.x`.
